File: src/extract_theme.py
#!/usr/bin/env python3
"""extract_theme.py

Extract color and font information from a PPTX file and generate a Marp theme (CSS).

A PPTX file is a ZIP archive. Its `ppt/theme/themeN.xml` entries define the color
scheme (<a:clrScheme>) and font scheme (<a:fontScheme>). This script extracts brand
colors and fonts from those entries and writes them into a CSS variable structure
compatible with src/themes/corporate.css.

Uses only the standard library (zipfile, xml.etree) — no external dependencies.

Usage:
    uv run python src/extract_theme.py data/pptx/sample.pptx
    uv run python src/extract_theme.py data/pptx/sample.pptx --name mybrand
    uv run python src/extract_theme.py data/pptx/sample.pptx -o src/themes/mybrand.css
"""

from __future__ import annotations

import argparse
import re
import sys
import zipfile
from pathlib import Path
from xml.etree import ElementTree as ET

# Fix encoding to UTF-8 on Windows consoles that default to cp932
for _stream in (sys.stdout, sys.stderr):
    try:
        _stream.reconfigure(encoding="utf-8")
    except (AttributeError, ValueError):
        pass

# DrawingML namespace
A = "{http://schemas.openxmlformats.org/drawingml/2006/main}"

ROOT = Path(__file__).resolve().parent.parent
DEFAULT_THEMES_DIR = ROOT / "src" / "themes"

# PPTX color scheme keys and their roles:
# dk1/lt1 = text/background 1, dk2/lt2 = text/background 2, accent1-6 = accents
COLOR_KEYS = [
    "dk1",
    "lt1",
    "dk2",
    "lt2",
    "accent1",
    "accent2",
    "accent3",
    "accent4",
    "accent5",
    "accent6",
    "hlink",
]

# Fallback values for Office system colors
SYS_COLOR_FALLBACK = {
    "windowText": "000000",
    "window": "FFFFFF",
}


def _read_theme_xml(pptx_path: Path) -> bytes:
    """Return the bytes of the first theme XML found inside the PPTX."""
    with zipfile.ZipFile(pptx_path) as z:
        theme_names = sorted(
            n for n in z.namelist() if re.match(r"ppt/theme/theme\d+\.xml$", n)
        )
        if not theme_names:
            raise ValueError("No theme XML found inside the PPTX.")
        return z.read(theme_names[0])


def _resolve_color(node: ET.Element) -> str | None:
    """Extract a 6-digit hex color from a color scheme node (<a:dk1>, etc.)."""
    srgb = node.find(f"{A}srgbClr")
    if srgb is not None and srgb.get("val"):
        return srgb.get("val").upper()
    sysclr = node.find(f"{A}sysClr")
    if sysclr is not None:
        # prefer lastClr attribute, fall back to known system color
        last = sysclr.get("lastClr")
        if last:
            return last.upper()
        return SYS_COLOR_FALLBACK.get(sysclr.get("val", ""), None)
    return None


def parse_theme(xml_bytes: bytes) -> dict:
    """Extract colors and fonts from theme XML bytes."""
    root = ET.fromstring(xml_bytes)

    colors: dict[str, str] = {}
    clr_scheme = root.find(f".//{A}clrScheme")
    if clr_scheme is not None:
        for key in COLOR_KEYS:
            node = clr_scheme.find(f"{A}{key}")
            if node is not None:
                hexval = _resolve_color(node)
                if hexval:
                    colors[key] = hexval

    fonts: dict[str, str] = {}
    font_scheme = root.find(f".//{A}fontScheme")
    if font_scheme is not None:
        major = font_scheme.find(f"{A}majorFont/{A}latin")
        minor = font_scheme.find(f"{A}minorFont/{A}latin")
        if major is not None and major.get("typeface"):
            fonts["major"] = major.get("typeface")
        if minor is not None and minor.get("typeface"):
            fonts["minor"] = minor.get("typeface")
        # pick up Japanese (East Asian) fonts if present
        major_ea = font_scheme.find(f'{A}majorFont/{A}font[@script="Jpan"]')
        minor_ea = font_scheme.find(f'{A}minorFont/{A}font[@script="Jpan"]')
        if major_ea is not None and major_ea.get("typeface"):
            fonts["major_ja"] = major_ea.get("typeface")
        if minor_ea is not None and minor_ea.get("typeface"):
            fonts["minor_ja"] = minor_ea.get("typeface")

    scheme_name = root.get("name", "extracted")
    return {"name": scheme_name, "colors": colors, "fonts": fonts}


def _hex(colors: dict, key: str, default: str) -> str:
    val = colors.get(key)
    return f"#{val}" if val else default


def _font_stack(fonts: dict, kind: str) -> str:
    """Build a CSS font stack from the major/minor font entries."""
    latin = fonts.get(kind)
    ja = fonts.get(f"{kind}_ja")
    parts = []
    if latin:
        parts.append(f'"{latin}"')
    if ja:
        parts.append(f'"{ja}"')
    # always append Japanese fallbacks
    parts += [
        '"Hiragino Kaku Gothic ProN"',
        '"Yu Gothic UI"',
        '"Meiryo"',
        "system-ui",
        "sans-serif",
    ]
    # deduplicate while preserving order
    seen = set()
    uniq = []
    for p in parts:
        if p not in seen:
            seen.add(p)
            uniq.append(p)
    return ", ".join(uniq)


def render_css(theme_name: str, parsed: dict) -> str:
    """Generate a Marp theme CSS compatible with corporate.css from extracted data."""
    colors = parsed["colors"]
    fonts = parsed["fonts"]

    primary = _hex(colors, "accent1", "#1f6feb")
    secondary = _hex(colors, "dk2", _hex(colors, "accent2", "#0b3d91"))
    accent = _hex(colors, "accent3", _hex(colors, "accent2", "#f0a202"))
    background = _hex(colors, "lt1", "#ffffff")
    foreground = _hex(colors, "dk1", "#22272e")
    link = _hex(colors, "hlink", primary)

    font_base = _font_stack(fonts, "minor")
    font_heading = _font_stack(fonts, "major")

    src_name = parsed.get("name", "extracted")
    return f"""/* @theme {theme_name} */

/*
 * Auto-generated by extract_theme.py from a PPTX file.
 * Source theme: {src_name}
 * Edit the :root variables below to customize colors and fonts.
 */

@import "default";

:root {{
  --brand-primary: {primary};
  --brand-secondary: {secondary};
  --brand-accent: {accent};

  --color-background: {background};
  --color-foreground: {foreground};
  --color-heading: var(--brand-secondary);
  --color-link: {link};

  --font-base: {font_base};
  --font-heading: {font_heading};
  --font-code: "Cascadia Code", "Consolas", "SFMono-Regular", monospace;

  font-family: var(--font-base);
  color: var(--color-foreground);
  background-color: var(--color-background);
}}

section {{
  font-family: var(--font-base);
  color: var(--color-foreground);
  background-color: var(--color-background);
  padding: 60px 70px;
  font-size: 26px;
  line-height: 1.5;
}}

h1, h2, h3, h4, h5, h6 {{
  font-family: var(--font-heading);
  color: var(--color-heading);
}}

h1 {{
  font-size: 1.9em;
  border-bottom: 4px solid var(--brand-primary);
  padding-bottom: 0.2em;
}}

h2 {{ font-size: 1.4em; }}

a {{ color: var(--color-link); }}
strong {{ color: var(--brand-secondary); }}
ul > li::marker {{ color: var(--brand-primary); }}

code {{ font-family: var(--font-code); }}
pre {{ border-radius: 8px; }}

table th {{
  background-color: var(--brand-primary);
  color: #ffffff;
}}

blockquote {{
  border-left: 6px solid var(--brand-accent);
  color: #4a5160;
}}

section::after {{
  color: #8b95a1;
  font-size: 0.6em;
}}

section.title {{
  justify-content: center;
  text-align: center;
  background: linear-gradient(135deg, var(--brand-secondary) 0%, var(--brand-primary) 100%);
  color: #ffffff;
}}
section.title h1 {{
  color: #ffffff;
  border-bottom: none;
  font-size: 2.4em;
}}
section.title h2, section.title h3 {{ color: rgba(255, 255, 255, 0.85); }}

section.section {{
  justify-content: center;
  background-color: var(--brand-secondary);
  color: #ffffff;
}}
section.section h1, section.section h2 {{
  color: #ffffff;
  border-bottom: none;
}}
"""


def slugify(text: str) -> str:
    s = re.sub(r"[^\w-]+", "-", text.strip().lower())
    s = re.sub(r"-+", "-", s).strip("-")
    return s or "extracted"


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(
        description="Extract a Marp theme (CSS) from a PPTX file"
    )
    parser.add_argument("pptx", help="path to the input PPTX file")
    parser.add_argument(
        "-n",
        "--name",
        help="theme name to use (@theme name; defaults to the PPTX filename)",
    )
    parser.add_argument(
        "-o",
        "--output",
        help="output CSS path (default: src/themes/<name>.css)",
    )
    parser.add_argument(
        "--print",
        action="store_true",
        help="print the CSS to stdout instead of writing a file",
    )
    args = parser.parse_args(argv)

    pptx_path = Path(args.pptx)
    if not pptx_path.exists():
        print(f"Error: file not found: {pptx_path}", file=sys.stderr)
        return 1

    try:
        xml_bytes = _read_theme_xml(pptx_path)
        parsed = parse_theme(xml_bytes)
    except (zipfile.BadZipFile, ValueError, ET.ParseError) as e:
        print(f"Error: failed to parse PPTX: {e}", file=sys.stderr)
        return 1

    theme_name = slugify(args.name) if args.name else slugify(pptx_path.stem)
    css = render_css(theme_name, parsed)

    if args.print:
        print(css)
        return 0

    out_path = (
        Path(args.output)
        if args.output
        else DEFAULT_THEMES_DIR / f"{theme_name}.css"
    )
    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_text(css, encoding="utf-8")

    print(f"[OK] Theme generated: {out_path}")
    print(f"  @theme name    : {theme_name}")
    print(f"  Source theme   : {parsed.get('name')}")
    if parsed["colors"]:
        cols = ", ".join(f"{k}=#{v}" for k, v in parsed["colors"].items())
        print(f"  Colors         : {cols}")
    if parsed["fonts"]:
        fns = ", ".join(f"{k}={v}" for k, v in parsed["fonts"].items())
        print(f"  Fonts          : {fns}")
    print()
    print(f"Usage: node src/convert.mjs --theme {theme_name} -f pdf")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: packages/core/src/extract-theme.mjs
/**
 * extract-theme.mjs
 * PPTX から配色・フォントを抽出し、Marp 用テーマ(CSS)を生成する。
 *
 * src/extract_theme.py(Python, 廃止)の JS 移植版。
 * PPTX は実体が ZIP。ppt/theme/themeN.xml の <a:clrScheme>/<a:fontScheme> 等を読む。
 * 依存は純 JS の fflate(unzip) + fast-xml-parser(XML) のみ。
 */
import { unzipSync, strFromU8 } from "fflate";
import { XMLParser } from "fast-xml-parser";

// 配色スキームのキー(dk1/lt1=テキスト/背景1, dk2/lt2=同2, accent1-6, hlink)
const COLOR_KEYS = [
  "dk1", "lt1", "dk2", "lt2",
  "accent1", "accent2", "accent3", "accent4", "accent5", "accent6",
  "hlink",
];

// Office 標準の system color フォールバック
const SYS_COLOR_FALLBACK = { windowText: "000000", window: "FFFFFF" };
const DEFAULT_SLIDE_SIZE_EMU = { slideWidthEmu: 12192000, slideHeightEmu: 6858000 };
const DEFAULT_CLR_MAP = {
  bg1: "lt1",
  tx1: "dk1",
  bg2: "lt2",
  tx2: "dk2",
  accent1: "accent1",
  accent2: "accent2",
  accent3: "accent3",
  accent4: "accent4",
  accent5: "accent5",
  accent6: "accent6",
  hlink: "hlink",
  folHlink: "folHlink",
};

const xmlParser = new XMLParser({
  ignoreAttributes: false,
  attributeNamePrefix: "@_",
  removeNSPrefix: false,
});

function toU8(pptxBytes) {
  return pptxBytes instanceof Uint8Array
    ? pptxBytes
    : new Uint8Array(pptxBytes instanceof ArrayBuffer ? pptxBytes : pptxBytes.buffer ?? pptxBytes);
}

function readPptxXmlParts(pptxBytes) {
  const files = unzipSync(toU8(pptxBytes), {
    filter: (f) =>
      /^ppt\/theme\/theme\d+\.xml$/.test(f.name) ||
      f.name === "ppt/presentation.xml" ||
      /^ppt\/slideMasters\/slideMaster\d+\.xml$/.test(f.name),
  });
  const themeNames = Object.keys(files).filter((name) => /^ppt\/theme\/theme\d+\.xml$/.test(name)).sort();
  const masterNames = Object.keys(files).filter((name) => /^ppt\/slideMasters\/slideMaster\d+\.xml$/.test(name)).sort();
  if (themeNames.length === 0) {
    throw new Error("PPTX 内に theme XML が見つかりませんでした。");
  }
  return {
    themeXml: strFromU8(files[themeNames[0]]),
    presentationXml: files["ppt/presentation.xml"] ? strFromU8(files["ppt/presentation.xml"]) : null,
    slideMasterXml: masterNames.length > 0 ? strFromU8(files[masterNames[0]]) : null,
  };
}

/** PPTX(Uint8Array/Buffer/ArrayBuffer) から最初のテーマ XML 文字列を取り出す。 */
function readThemeXml(pptxBytes) {
  return readPptxXmlParts(pptxBytes).themeXml;
}

/** 配色ノード(dk1 等の中身)から 6桁 HEX を取り出す。 */
function resolveColor(node) {
  if (!node || typeof node !== "object") return null;
  const srgb = node["a:srgbClr"];
  if (srgb && srgb["@_val"]) return String(srgb["@_val"]).toUpperCase();
  const sys = node["a:sysClr"];
  if (sys) {
    if (sys["@_lastClr"]) return String(sys["@_lastClr"]).toUpperCase();
    return SYS_COLOR_FALLBACK[sys["@_val"]] ?? null;
  }
  return null;
}

/** テーマ XML 文字列から色とフォントを抽出する。 */
export function parseThemeXml(xml) {
  const doc = xmlParser.parse(xml);
  const theme = doc["a:theme"] || {};
  const elements = theme["a:themeElements"] || {};

  const colors = {};
  const clrScheme = elements["a:clrScheme"];
  if (clrScheme) {
    for (const key of COLOR_KEYS) {
      const hex = resolveColor(clrScheme[`a:${key}`]);
      if (hex) colors[key] = hex;
    }
  }

  const fonts = {};
  const fontScheme = elements["a:fontScheme"];
  if (fontScheme) {
    const major = fontScheme["a:majorFont"];
    const minor = fontScheme["a:minorFont"];
    const latin = (f) => f?.["a:latin"]?.["@_typeface"] || null;
    const jpan = (f) => {
      const list = f?.["a:font"];
      const arr = Array.isArray(list) ? list : list ? [list] : [];
      const hit = arr.find((x) => x["@_script"] === "Jpan");
      return hit?.["@_typeface"] || null;
    };
    if (latin(major)) fonts.major = latin(major);
    if (latin(minor)) fonts.minor = latin(minor);
    if (jpan(major)) fonts.major_ja = jpan(major);
    if (jpan(minor)) fonts.minor_ja = jpan(minor);
  }

  return { name: theme["@_name"] || "extracted", colors, fonts };
}

/** presentation.xml からスライドサイズ(EMU)を抽出する。 */
export function parsePresentation(xml) {
  if (!xml) return { ...DEFAULT_SLIDE_SIZE_EMU };
  const doc = xmlParser.parse(xml);
  const sldSz = doc?.["p:presentation"]?.["p:sldSz"];
  const width = Number(sldSz?.["@_cx"]);
  const height = Number(sldSz?.["@_cy"]);
  return {
    slideWidthEmu: Number.isFinite(width) && width > 0 ? width : DEFAULT_SLIDE_SIZE_EMU.slideWidthEmu,
    slideHeightEmu: Number.isFinite(height) && height > 0 ? height : DEFAULT_SLIDE_SIZE_EMU.slideHeightEmu,
  };
}

function firstNode(node) {
  return Array.isArray(node) ? node[0] : node;
}

function hundredthsToPt(value) {
  const n = Number(value);
  return Number.isFinite(n) && n > 0 ? n / 100 : null;
}

function normalizeHex(hexValue) {
  if (!hexValue) return null;
  const hexText = String(hexValue).replace(/^#/, "").toUpperCase();
  return /^[0-9A-F]{6}$/.test(hexText) ? hexText : null;
}

function parseClrMap(master) {
  const attrs = master?.["p:clrMap"];
  if (!attrs || typeof attrs !== "object") return { ...DEFAULT_CLR_MAP };
  const normalized = {};
  for (const [key, value] of Object.entries(attrs)) {
    normalized[key.replace(/^@_/, "")] = value;
  }
  return { ...DEFAULT_CLR_MAP, ...normalized };
}

/** schemeClr token を clrMap 経由で clrScheme の 6桁 HEX に解決する。 */
export function resolveSchemeColor(token, clrMap = DEFAULT_CLR_MAP, clrScheme = {}) {
  const mapped = clrMap?.[token] || token;
  return normalizeHex(clrScheme?.[mapped]) || normalizeHex(clrScheme?.[token]) || null;
}

function resolveSolidFillColor(solidFill, clrMap, clrScheme) {
  if (!solidFill) return null;
  const direct = resolveColor(solidFill);
  if (direct) return normalizeHex(direct);
  const schemeToken = solidFill["a:schemeClr"]?.["@_val"];
  return schemeToken ? resolveSchemeColor(schemeToken, clrMap, clrScheme) : null;
}

function parseTextStyle(style) {
  const lvl1 = firstNode(style?.["a:lvl1pPr"]);
  return {
    pt: hundredthsToPt(lvl1?.["a:defRPr"]?.["@_sz"]),
    align: lvl1?.["@_algn"] || null,
  };
}

function normalizeAlign(value) {
  if (value === "ctr") return "center";
  if (value === "l") return "left";
  if (value === "r") return "right";
  return value || null;
}

/** slideMaster XML からタイトル/本文サイズ、タイトル揃え、背景色を抽出する。 */
export function parseSlideMaster(xml, clrScheme = {}) {
  if (!xml) return { titlePt: null, bodyPt: null, titleAlign: null, bgColorHex: null };
  const doc = xmlParser.parse(xml);
  const master = doc?.["p:sldMaster"] || {};
  const clrMap = parseClrMap(master);
  const bg = master?.["p:cSld"]?.["p:bg"];
  const bgPr = bg?.["p:bgPr"];
  const bgRef = bg?.["p:bgRef"];
  const bgColorHex =
    resolveSolidFillColor(bgPr?.["a:solidFill"], clrMap, clrScheme) ||
    (bgRef?.["a:schemeClr"]?.["@_val"]
      ? resolveSchemeColor(bgRef["a:schemeClr"]["@_val"], clrMap, clrScheme)
      : null);

  const txStyles = master?.["p:txStyles"] || {};
  const title = parseTextStyle(txStyles["p:titleStyle"]);
  const body = parseTextStyle(txStyles["p:bodyStyle"]);

  return {
    titlePt: title.pt,
    bodyPt: body.pt,
    titleAlign: normalizeAlign(title.align),
    bgColorHex,
  };
}

function hex(colors, key, fallback) {
  return colors[key] ? `#${colors[key]}` : fallback;
}

/** major/minor を CSS フォントスタックへ。日本語フォールバックを必ず付与。 */
function fontStack(fonts, kind) {
  const latin = fonts[kind];
  const ja = fonts[`${kind}_ja`];
  const parts = [];
  if (latin) parts.push(`"${latin}"`);
  if (ja) parts.push(`"${ja}"`);
  parts.push(
    '"Hiragino Kaku Gothic ProN"', '"Yu Gothic UI"', '"Meiryo"', "system-ui", "sans-serif"
  );
  return [...new Set(parts)].join(", ");
}

/** kebab-case 化。 */
export function slugify(text) {
  const raw = String(text ?? "").trim().toLowerCase();
  if (!raw) return "extracted";
  const s = raw.normalize("NFKD").replace(/[\u0300-\u036f]/g, "").replace(/[^a-z0-9-]+/g, "-").replace(/-+/g, "-").replace(/^-|-$/g, "");
  if (s) return s;
  const hash = [...raw].reduce((acc, ch) => (acc * 31 + ch.codePointAt(0)) >>> 0, 0).toString(36);
  return `extracted-${hash}`;
}

/** 抽出結果から corporate.css 互換の Marp テーマ CSS を生成する。 */
export function renderThemeCss(themeName, parsed) {
  const { colors = {}, fonts = {}, sizes = {}, background } = parsed;
  const primary = hex(colors, "accent1", "#1f6feb");
  const secondary = hex(colors, "dk2", hex(colors, "accent2", "#0b3d91"));
  const accent = hex(colors, "accent3", hex(colors, "accent2", "#f0a202"));
  const backgroundHex = normalizeHex(background);
  const backgroundColor = backgroundHex ? `#${backgroundHex}` : hex(colors, "lt1", "#ffffff");
  const foreground = hex(colors, "dk1", "#22272e");
  const link = hex(colors, "hlink", primary);
  const fontBase = fontStack(fonts, "minor");
  const fontHeading = fontStack(fonts, "major");
  const srcName = parsed.name || "extracted";
  const pxPerPt = sizes.pxPerPt ?? (1280 / (DEFAULT_SLIDE_SIZE_EMU.slideWidthEmu / 12700));
  const bodyPt = sizes.bodyPt ?? 24;
  const titlePt = sizes.titlePt ?? 40;
  const bodyPx = Math.round(bodyPt * pxPerPt);
  const titlePx = Math.round(titlePt * pxPerPt);
  const titleRatio = titlePx / bodyPx;
  const extractedDetails = [
    sizes.titlePt != null ? `タイトル: ${sizes.titlePt}pt` : null,
    sizes.bodyPt != null ? `本文: ${sizes.bodyPt}pt` : null,
    backgroundHex ? `背景: #${backgroundHex}` : null,
  ].filter(Boolean).join(" / ");

  return `/* @theme ${themeName} */

/*
 * このテーマは extract-theme により PPTX から自動生成されました。
 * 元テーマ名: ${srcName}
${extractedDetails ? ` * 抽出情報: ${extractedDetails}\n` : ""} * 色やフォントは下記 :root の変数を編集して微調整できます。
 */

@import "default";

:root {
  --brand-primary: ${primary};
  --brand-secondary: ${secondary};
  --brand-accent: ${accent};

  --color-background: ${backgroundColor};
  --color-foreground: ${foreground};
  --color-heading: var(--brand-secondary);
  --color-link: ${link};

  --font-base: ${fontBase};
  --font-heading: ${fontHeading};
  --font-code: "Cascadia Code", "Consolas", "SFMono-Regular", monospace;

  font-family: var(--font-base);
  color: var(--color-foreground);
  background-color: var(--color-background);
}

section {
  font-family: var(--font-base);
  color: var(--color-foreground);
  background-color: var(--color-background);
  padding: 60px 70px;
  font-size: ${bodyPx}px;
  line-height: 1.5;
}

h1, h2, h3, h4, h5, h6 {
  font-family: var(--font-heading);
  color: var(--color-heading);
}

h1 {
  font-size: ${titleRatio.toFixed(2)}em;
  border-bottom: 4px solid var(--brand-primary);
  padding-bottom: 0.2em;
}

h2 { font-size: 1.4em; }

a { color: var(--color-link); }
strong { color: var(--brand-secondary); }
ul > li::marker { color: var(--brand-primary); }

code { font-family: var(--font-code); }
pre { border-radius: 8px; }

table th {
  background-color: var(--brand-primary);
  color: #ffffff;
}

blockquote {
  border-left: 6px solid var(--brand-accent);
  color: #4a5160;
}

section::after {
  color: #8b95a1;
  font-size: 0.6em;
}

section.title {
  justify-content: center;
  text-align: center;
  background: linear-gradient(135deg, var(--brand-secondary) 0%, var(--brand-primary) 100%);
  color: #ffffff;
}
section.title h1 {
  color: #ffffff;
  border-bottom: none;
  font-size: ${(titleRatio * 1.4).toFixed(2)}em;
}
section.title h2, section.title h3 { color: rgba(255, 255, 255, 0.85); }

section.section {
  justify-content: center;
  background-color: var(--brand-secondary);
  color: #ffffff;
}
section.section h1, section.section h2 {
  color: #ffffff;
  border-bottom: none;
}
`;
}

/**
 * PPTX バイト列からテーマ CSS を生成する高レベル API。
 * @param {Uint8Array|ArrayBuffer|Buffer} pptxBytes
 * @param {object} [opts]
 * @param {string} [opts.name] テーマ名(@theme 名)。省略時は元テーマ名を slug 化。
 * @returns {{ css:string, themeName:string, parsed:object }}
 */
export function extractThemeFromPptx(pptxBytes, { name } = {}) {
  const parts = readPptxXmlParts(pptxBytes);
  const parsed = parseThemeXml(parts.themeXml);
  const presentation = parsePresentation(parts.presentationXml);
  const master = parseSlideMaster(parts.slideMasterXml, parsed.colors);
  const slideWidthPt = presentation.slideWidthEmu / 12700;
  const pxPerPt = 1280 / slideWidthPt;
  parsed.presentation = presentation;
  parsed.sizes = {
    titlePt: master.titlePt,
    bodyPt: master.bodyPt,
    pxPerPt,
  };
  parsed.titleAlign = master.titleAlign;
  parsed.background = master.bgColorHex;
  const themeName = slugify(name || parsed.name);
  const css = renderThemeCss(themeName, parsed);
  return { css, themeName, parsed };
}
